### ISSUE TITLE
[C++] Capture/save all threaded exceptions and throw a 'meta' exception

### ISSUE THREAD

**Issue by suntzu86** [Yelp/MOE#249]
BLOCKED BY: #210 
This is a follow-up ticket.

From discussion with @ypwais here: https://github.com/sc932/MOE/pull/238.
@ypwais  wrote

```
The only problematic consequence of this design that I see right now is that returning only the first exception could mask other errors (e.g. multiple threads raise different kinds of exceptions)

...

One strategy here is to re-throw a runtime error that includes all the exceptions encountered, i.e. some message that includes all the different tracebacks. It sounds like that info would probably not be super useful at the moment, but perhaps more relevant if this part of the code moved to a more complicated concurrency model (e.g. adding retries). So ya I vote fine to leave for now; I mainly brought up the issue to offer some discussion on probably the main core decision made in this change.
```

Right now, I will rethrow the first (temporally) exception caught in threading. Other exceptions are logged alongside the thread id and iteration number (so the info is available but not as easily accessed and not programmatically).
